mono/metadata/threads.c

```diff
@@ -4678,12 +4660,25 @@ static void CALLBACK dummy_apc (ULONG_PTR param)
  * Performs the operation that the requested thread state requires (abort,
  * suspend or stop)
  */
-static MonoExceptionHandle
-mono_thread_execute_interruption (void)
+static gboolean
+mono_thread_execute_interruption (MonoExceptionHandle *pexc)
 {
-	MONO_REQ_GC_UNSAFE_MODE;
+	// Use caller's frame if provided, else create one.
+	// This is not just an optimization -- it is needed to return
+	// a handle to the caller through an out-parameter.
+	//
+	// Caller cannot do the handle allocation because it has loops.
```

Comment:
What am  I missing? Why can't the caller allocate a handle outside the loop?
```
   MonoExceptionHandle exc = MONO_HANDLE_NEW (..., NULL);
   for (;;) {
     ...
     if (mono_thread_execute_interruption (&exc)) {
       mono_error_set_exception_handle (error, exc);
       return
     }
  }
```

Is there some reason why this wouldn't work?